cirq/contrib/qasm_import/_parser.py

```diff
@@ -35,7 +36,42 @@ def __init__(self, supported_format: bool, qelib1_include: bool,
         self.circuit = c
 
 
-class QasmParser(object):
+class QasmGate:
+
+    def __init__(self, qasm_gate: str, cirq_gate: cirq.Gate, num_args: int):
+        self.qasm_gate = qasm_gate
+        self.cirq_gate = cirq_gate
+        self.num_args = num_args
+
+    def validate_args(self, args, lineno):
+        if len(args) != self.num_args:
+            raise QasmException(
+                "{} only takes {} arg(s) (qubits and/or registers), "
+                "got: {}, at line {}".format(self.qasm_gate, self.num_args,
+                                             len(args), lineno))
+
+    def on(self, args: List[List[cirq.Qid]],
+           lineno: int = 0) -> Iterable[cirq.Operation]:
+        self.validate_args(args, lineno)
+        reg_size = 1
+        for reg in args:
+            if len(reg) > 1 and len(reg) != reg_size:
+                if reg_size == 1:
+                    reg_size = len(reg)
+                else:
+                    raise QasmException(
+                        "Non matching quantum registers of length {} "
+                        "at line {}".format([len(reg) for reg in args], lineno))
+
+        for i in range(reg_size):
+            # because of mixed quantum register/qubit support there can
+            # be single qubit registers and longer qubit registers
+            # for single qubit registers qreg[0], for longer ones qreg[i]
```

Comment:
I don't understand what this comment is saying and how it relates to the code below.

Reply:
Yeah this could be clearer. 

Maybe: 
```
# OpenQASM gates can be applied on single qubits and qubit registers. 
# We represent single qubits as registers of size 1. 
# Based on the OpenQASM spec (https://arxiv.org/abs/1707.03429), 
# single qubit arguments can be mixed with qubit registers.
# Given quantum registers of length n and single qubits are both used as arguments,
# we generate n `GateOperations` via iterating through each qubit of the registers
# 0 to n-1 and use the same one qubit from the "single-qubit registers" for each operation. 
```

Let's look at the example of two 2-qubit registers: 

```
qreg q1[2];
qreg q2[2]; 
```

We can do the following statements: 

```
CX q1, q2; 
```
Which means
```
         ┌──┐
q1_0: ────@─────
          │
q1_1: ────┼@────
          ││
q2_0: ────X┼────
           │
q2_1: ─────X────
         └──┘
```

While a single qubit mixed with a register: 
```
cx q1[0], q2;
```
Means:
```
q1_0: ───@───@───
         │   │
q2_0: ───X───┼───
             │
q2_1: ───────X───
```

Reply:
At this point I might refactor this to make it more understandable though.

Reply:
Also TODO for a new validation: the args shouldn't overlap in qubits ever i.e. these are invalid: 

```
CX q1[0], q1;
CX q2[1], q2[1];
CX q1, q1;
```

Reply:
I would describe this by talking about how we need to implement "broadcast semantics", which is a common thing from numpy.

Reply:
Interestingly cirq already has a validation here. I wonder if I should just catch errors on operation creation and throw expose them in the parser nicely so that user can see the line where it was thrown or should I write my own overlap checking logic here...

```
ValueError: Overlapping operations: (cirq.CNOT.on(cirq.NamedQubit('q1_1'), cirq.NamedQubit('q1_1')),)
```